### ISSUE TITLE
question about the pos_neg

### ISSUE THREAD

**Issue by yex1a0** [ModelTC/United-Perception#73]
I tried to apply your efl loss to a DETR-based detector, where the method for collecting gradients was the manual collection method you mentioned in the issue.https://github.com/ModelTC/United-Perception/issues/59#issuecomment-1409673016
 However, I noticed that the pos_neg parameter is the ratio of all positive sample gradients to negative sample gradients. For some reason, the positive sample gradients were much larger than the negative sample gradients in the early stage, resulting in the accumulated positive sample gradients always being larger than the negative sample gradients to a certain extent, causing the pos_neg parameter to be 1 for all categories, whether head or tail, degenerating into a normal focal loss. Why is this? How can I solve this problem?
my code:
import torch
from .efl_utils import GeneralizedCrossEntropyLoss,_reduce,allreduce

class EqualizedFocalLoss(GeneralizedCrossEntropyLoss):
    def __init__(self,
                 name='equalized_focal_loss',
                 reduction='mean',
                 loss_weight=1.0,
                 ignore_index=-1,
                 num_classes=1204,
                 focal_gamma=2.0,
                 focal_alpha=0.25,
                 scale_factor=8.0,
                 fpn_levels=5):
        activation_type = 'sigmoid'
        GeneralizedCrossEntropyLoss.__init__(self,
                                             name=name,
                                             reduction=reduction,
                                             loss_weight=loss_weight,
                                             activation_type=activation_type,
                                             ignore_index=ignore_index)

        # cfg for focal loss
        self.focal_gamma = focal_gamma
        self.focal_alpha = focal_alpha

        # ignore bg class and ignore idx
        self.num_classes = num_classes - 1

        # cfg for efl loss
        self.scale_factor = scale_factor
        # initial variables
        self.register_buffer('pos_grad', torch.zeros(self.num_classes))
        self.register_buffer('neg_grad', torch.zeros(self.num_classes))
        self.register_buffer('pos_neg', torch.ones(self.num_classes))

        # grad collect
        self.grad_buffer = []
        self.fpn_levels = fpn_levels

    def forward(self, input, target, reduction, normalizer=None):
        self.n_c = input.shape[-1]
        self.input = input.reshape(-1, self.n_c)
        self.target = target.reshape(-1)
        self.n_i, _ = self.input.size()

        def expand_label(pred, gt_classes):
            target = pred.new_zeros(self.n_i, self.n_c + 1)
            target[torch.arange(self.n_i), gt_classes] = 1
            return target[:, :-1]

        expand_target = expand_label(self.input, self.target)
        sample_mask = (self.target != self.ignore_index)

        inputs = self.input[sample_mask]
        targets = expand_target[sample_mask]
        self.cache_mask = sample_mask
        self.cache_target = expand_target

        pred = torch.sigmoid(inputs)
        pred_t = pred * targets + (1 - pred) * (1 - targets)
        map_val = 1 - self.pos_neg.detach()
        dy_gamma = self.focal_gamma + self.scale_factor * map_val
        # focusing factor
        ff = dy_gamma.view(1, -1).expand(self.n_i, self.n_c)[sample_mask]
        # weighting factor
        wf = ff / self.focal_gamma

        # ce_loss
        ce_loss = -torch.log(pred_t)
        cls_loss = ce_loss * torch.pow((1 - pred_t), ff.detach()) * wf.detach()

        # to avoid an OOM error
        torch.cuda.empty_cache()

        if self.focal_alpha >= 0:
            alpha_t = self.focal_alpha * targets + (1 - self.focal_alpha) * (1 - targets)
            cls_loss = alpha_t * cls_loss

        self.collect_grad_manual(inputs, targets, ff.detach(), normalizer or 1.0)

        if normalizer is None:
            normalizer = 1.0

        return _reduce(cls_loss, reduction, normalizer=normalizer)

    def collect_grad(self, grad_in):
        bs = grad_in.shape[0]
        self.grad_buffer.append(grad_in.detach().permute(0, 2, 3, 1).reshape(bs, -1, self.num_classes))
        if len(self.grad_buffer) == self.fpn_levels:
            target = self.cache_target[self.cache_mask]
            grad = torch.cat(self.grad_buffer[::-1], dim=1).reshape(-1, self.num_classes)

            grad = torch.abs(grad)[self.cache_mask]
            pos_grad = torch.sum(grad * target, dim=0)
            neg_grad = torch.sum(grad * (1 - target), dim=0)

            allreduce(pos_grad)
            allreduce(neg_grad)

            self.pos_grad += pos_grad
            self.neg_grad += neg_grad
            self.pos_neg = torch.clamp(self.pos_grad / (self.neg_grad + 1e-10), min=0, max=1)

            self.grad_buffer = []
    
    def collect_grad_manual(self, input, target, focus_factor, normalizer):
        prob = torch.sigmoid(input)
        pos_grad = target * torch.pow((1 - prob), focus_factor) * (focus_factor * prob * torch.log(prob) + prob - 1) * focus_factor / self.focal_gamma
        neg_grad = (target - 1) * torch.pow(prob, focus_factor) * (focus_factor * (1 - prob) * torch.log(1 - prob) - prob) * focus_factor / self.focal_gamma
        if self.focal_alpha >= 0:
            pos_grad = pos_grad * self.focal_alpha
            neg_grad = neg_grad * (1 - self.focal_alpha)
        grad = pos_grad + neg_grad
        grad = torch.abs(grad / normalizer)
        pos_grad = torch.sum(grad * target, dim=0)
        neg_grad = torch.sum(grad * (1 - target), dim=0)
        allreduce(pos_grad)
        allreduce(neg_grad)
        self.pos_grad += pos_grad
        self.neg_grad += neg_grad
        self.pos_neg = torch.clamp(self.pos_grad / (self.neg_grad + 1e-10), min=0, max=1)
![Image](https://github.com/user-attachments/assets/65c13659-ee8a-4fd0-898c-e06b79d5f632)
